docs/networking/streamcodecs.md

```diff
@@ -0,0 +1,369 @@
+
+`ByteBufCodecs#idMapper` provides a convenient way to send ids for objects. It either takes in two functions which convert an object to int and vice versa, or an `IdMap`.
+
+```java
+// For some enum
+public enum ExampleIdObject {
+    ;
+
+    // Gets Id -> Enum
+    public static final IntFunction<ExampleIdObject> BY_ID = 
+        ByIdMap.continuous(
+            ExampleIdObject::getId,
+            ExampleIdObject.values(),
+        ByIdMap.OutOfBoundsStrategy.ZERO
+    );
+    
+    ExampleIdObject(int id) { /* ... */ }
+}
+
+// The stream codec would look like
+public static final StreamCodec<ByteBuf, ExampleIdObject> ID_STREAM_CODEC =
+    ByteBufCodecs.idMapper(ExampleIdObject.BY_ID, ExampleIdObject::getId);
+```
+
+:::note
+NeoForge provides an alternative for id mappers that does not cache the enum values on construction via `IExtensibleEnum#createStreamCodecForExtensibleEnum`. However, this rarely needs to be used outside of extensible enums.
+:::
+
+### Optional
+
+A stream codec for sending an `Optional` wrapped value can be generated by supplying a stream codec to `ByteBufCodecs#optional`. This method first reads/writes a boolean indicating whether to read/write the object.
+
+```java
+public static final StreamCodec<RegistryFriendlyByteBuf, Optional<DataComponentType<?>>> OPTIONAL_STREAM_CODEC =
+    DataComponentType.STREAM_CODEC.apply(ByteBufCodecs::optional);
+```
+
+### Registry Objects
+
+Registry objects can be sent across the network using one of three methods: `registry`, `holderRegistry`, or `holder`. Each takes in a `ResourceKey` representing the registry the registry object is in.
+
+`registry` and `holderRegistry` returns the registry object or a holder wrapped registry object, respectively. These methods send over an id representing the registry object.
+
+```java
+// Registry object
+public static final StreamCodec<RegistryFriendlyByteBuf, Item> VALUE_STREAM_CODEC =
+    BytebufCodecs.registry(Registries.ITEM);
+
+// Holder of registry object
+public static final StreamCodec<RegistryFriendlyByteBuf, Holder<Item>> HOLDER_STREAM_CODEC =
+    BytebufCodecs.holderRegistry(Registries.ITEM);
+```
+
+`holder` returns a holder wrapped registry object. This method sends over an id representing the registry object, or the registry object itself if the provided `Holder` is a direct reference. To do so, `holder` also takes in the stream codec of the registry object.
+
+```java
+public static final StreamCodec<RegistryFriendlyByteBuf, Holder<SoundEvent>> STREAM_CODEC =
+    ByteBufCodecs.holder(
+        Registries.SOUND_EVENT, SoundEvent.DIRECT_STREAM_CODEC
+    );
+```
+
+### Holder Sets
+
+Tags or sets of holder wrapped registry objects can be sent using `holderSet`. This takes in a `ResourceKey` represent the registry the registry objects are in.
```

Comment:
```suggestion
Tags or sets of holder wrapped registry objects can be sent using `holderSet`. This takes in a `ResourceKey` representing the registry the registry objects are in.
```

This should mention that for actual tags the tag name is sent and for non-tag holder sets a `StreamCodec` constructed through `holderRegistry()` is used. This also means that the same rules regarding synced registries apply.